anesthetic/utils.py

```diff
@@ -182,3 +184,132 @@ def iso_probability_contours(pdf, contours=[0.68, 0.95]):
                 c[j] = c[j] - 1e-5
 
     return c
+
+
+def iso_probability_contours_from_samples(pdf, contours=[0.68, 0.95],
+                                          weights=None):
+    """Compute the iso-probability contour values."""
+    if weights is None:
+        weights = numpy.ones_like(pdf)
+    contours = [1-p for p in reversed(contours)]
+    i = numpy.argsort(pdf)
+    m = numpy.cumsum(weights[i])
+    m /= m[-1]
+    interp = interp1d([0]+list(m), [0]+list(pdf[i]))
+    c = list(interp(contours))+[max(pdf)]
+
+    # Correct level sets
+    for i in range(1, len(c)):
+        if c[i-1] == c[i]:
+            for j in range(i):
+                c[j] = c[j] - 1e-5
+
+    return c
+
+
+def triangular_sample_compression_2d(x, y, w=None, n=1000):
+    """Histogram a 2D set of weighted samples via triangulation.
+
+    This defines bins via a triangulation of the subsamples, sums weights
+    within triangles, and computes weighted centroids of triangles.
+
+    Parameters
+    ----------
+    x, y: array-like
+        x and y coordinates of samples for compressing
+
+    w: array-like, optional
+        weights of samples
+
+    n: int, optional
+        number of samples returned. Default 1000
+
+    Returns
+    -------
+    x, y, w, array-like
+        Compressed samples and weights
+
+    """
+    if w is None:
+        w = numpy.ones_like(x)
+
+    # Select samples for triangulation
+    if sum(w != 0) < n:
+        i = numpy.arange(len(w))
+    else:
+        i = numpy.random.choice(len(w), size=n, replace=False, p=w/w.sum())
+
+    # Generate triangulation
+    cov = numpy.cov(x, y, aweights=w)
+    L = numpy.linalg.cholesky(cov)
+    Linv = numpy.linalg.inv(L)
+    x_, y_ = Linv.dot([x[i], y[i]])
+    tri = Triangulation(x_, y_)
+
+    # Mask out triangles with unreasonably large perimeters
+    vec = numpy.array([x_[tri.triangles], y_[tri.triangles]]).transpose()
+    s = (numpy.linalg.norm(vec[1, :, :] - vec[0, :, :], axis=1) +
+         numpy.linalg.norm(vec[2, :, :] - vec[1, :, :], axis=1) +
+         numpy.linalg.norm(vec[0, :, :] - vec[2, :, :], axis=1))
+
+    # Mask out triangles with a perimeter zscore smaller than expected
+    tri.set_mask(zscore(numpy.log(s)) > -norm.ppf(1/len(s)))
```

Comment:
Have you looked into [`matplotlib.tri.TriAnalyzer`](https://matplotlib.org/3.1.1/api/tri_api.html#matplotlib.tri.TriAnalyzer) and in particular [`get_flat_tri_mask`](https://matplotlib.org/3.1.1/api/tri_api.html#matplotlib.tri.TriAnalyzer.get_flat_tri_mask)? Could this be even better than our manual masking?

Reply:
>Have you looked into matplotlib.tri.TriAnalyzer and in particular get_flat_tri_mask? Could this be even better than our manual masking?

I did try this first. The trouble is that it's not just flat triangles we want to be rid of, but over-large ones as well.